### PR TITLE
Add stub for AsyncExitContext added by bpo-29302.

### PR DESCRIPTION
**Pull request by Kentzo**
Stub for changes added to Python 3.7 in https://github.com/python/cpython/pull/4790.

### CODE REVIEW

File: stdlib/2and3/contextlib.pyi [hunk 1]
```diff
@@ -64,3 +64,29 @@ if sys.version_info >= (3,):
         def pop_all(self) -> ExitStack: ...
         def close(self) -> None: ...
         def __enter__(self: _U) -> _U: ...
+
+if sys.version_info >= (3, 7):
+    from asyncio.futures import Future
```
Comment (JelleZijlstra) [inline stdlib/2and3/contextlib.pyi]: Why not use typing.Awaitable instead of Future?

Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: I was looking at stubs for `asyncio.tasks` as an example.

File: stdlib/2and3/contextlib.pyi [hunk 2]
```diff
@@ -64,3 +64,29 @@ if sys.version_info >= (3,):
...
+        def __init__(self) -> None: ...
+        def enter_context(self, cm: ContextManager[_T]) -> _T: ...
+        def enter_async_context(self, cm: AsyncContextManager[_T]) -> Awaitable[_T]: ...
+        def push(self, exit: _CM_EF) -> _CM_EF: ...
+        def push_async_exit(self, exit: _ACM_EF) -> _ACM_EF: ...
+        def callback(self, callback: Callable[..., None],
+                     *args: Any, **kwds: Any) -> Callable[..., None]: ...
+        def push_async_callback(self, callback: _CallbackCoroFunc,
+                                *args: Any, **kwds: Any) -> _CallbackCoroFunc: ...
+        def pop_all(self) -> ExitStack: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/contextlib.pyi]: Should return AsyncExitStack.

File: stdlib/2and3/contextlib.pyi [hunk 3]
```diff
@@ -64,3 +64,29 @@ if sys.version_info >= (3,):
...
+        def aclose(self) -> Awaitable[None]: ...
+        def __enter__(self: _U) -> _U: ...
+        def __aenter__(self: _U) -> Awaitable[_U]: ...
```
Comment (JelleZijlstra) [inline stdlib/2and3/contextlib.pyi]: Do you not need `__exit__` and `__aexit__`? I suppose the base class provides those.

Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: Base class implementations should be enough.

File: stdlib/2and3/contextlib.pyi (hunk 2 above)
Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: Actually, I think it should be implemented as `__enter__` / `__aenter__`: `pop_all` [instantiates](https://github.com/python/cpython/blob/master/Lib/contextlib.py#L391) `type(self)`.

Probably that should also be fixed for `ExitStack`.

Comment (JelleZijlstra) [reply stdlib/2and3/contextlib.pyi]: Thanks! Would you mind fixing it there too while you're at it?

File: stdlib/2and3/contextlib.pyi (hunk 3 above)
Comment (JelleZijlstra) [reply stdlib/2and3/contextlib.pyi]: Actually, that's true for `__aexit__` but not `__exit__`.

Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: Why not? Definitions in typing.pyi seem legit ([`__exit__`](https://github.com/python/typeshed/blob/71c2a94cb896b2b121b53ab66777a9819265f00e/stdlib/3/typing.pyi#L317-L319) / [`__aexit__`](https://github.com/python/typeshed/blob/71c2a94cb896b2b121b53ab66777a9819265f00e/stdlib/3/typing.pyi#L325-L327))

Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: Or do you mean that's because `AsyncContextManager` does not inherit from `ContextManager`?

Comment (Kentzo) [reply stdlib/2and3/contextlib.pyi]: Oh, nvm. There is not `__enter__` for `AsyncExitStack`.

**Comment by Kentzo** [comment]
@JelleZijlstra Should I squash into 2 commits: one for AsyncExitStack and one for ExitStack's bugfix?

**Comment by JelleZijlstra** [comment]
No, that's fine. I'll squash when I merge (which I will do as soon as CI passes).

**Pull request merged**